File: coil_js/src/compare.js
// compare.js — semantic lossless validator

function extractRows(obj) {
  if (Array.isArray(obj)) return obj;

  if (obj && typeof obj === "object") {
    for (const k in obj) {
      if (Array.isArray(obj[k])) return obj[k];
    }
  }
  return [];
}

function canonicalRow(row) {
  if (typeof row !== "object" || row === null) return String(row);

  return Object.keys(row)
    .sort()
    .map(k => `${k}:${String(row[k])}`)
    .join("|");
}

function isLossless(original, decoded) {
  const oRows = extractRows(original);
  const dRows = extractRows(decoded);

  if (!oRows.length || !dRows.length) return false;

  const count = rows => {
    const map = {};
    rows.forEach(r => {
      const key = canonicalRow(r);
      map[key] = (map[key] || 0) + 1;
    });
    return map;
  };

  const oMap = count(oRows);
  const dMap = count(dRows);

  if (Object.keys(oMap).length !== Object.keys(dMap).length) return false;

  for (const k in oMap) {
    if (oMap[k] !== dMap[k]) return false;
  }

  return true;
}

module.exports = { isLossless };

File: compare.py
import json
from collections import Counter


def extract_rows(obj):
    """
    Extract all row-like structures regardless of nesting.
    """
    if isinstance(obj, list):
        return obj

    if isinstance(obj, dict):
        for v in obj.values():
            if isinstance(v, list):
                return v

    return []


def canonical_row(row):
    """
    Convert row to comparable form.
    """
    if not isinstance(row, dict):
        return str(row)

    return tuple(
        (k, str(row[k]))
        for k in sorted(row.keys())
    )


def isLossless(original, decoded):
    """
    True semantic comparison (order-independent).
    """

    o_rows = extract_rows(original)
    d_rows = extract_rows(decoded)

    if not o_rows or not d_rows:
        return False

    o_set = Counter(canonical_row(r) for r in o_rows)
    d_set = Counter(canonical_row(r) for r in d_rows)

    return o_set == d_set
